integration/container/health_test.go

```diff
@@ -93,6 +94,44 @@ while true; do sleep 1; done
 	poll.WaitOn(t, pollForHealthStatus(ctxPoll, client, id, "healthy"), poll.WithDelay(100*time.Millisecond))
 }
 
+// TestHealthCheckProcessKilled verifies that health-checks exec get killed on time-out.
+func TestHealthCheckProcessKilled(t *testing.T) {
+	defer setupTest(t)()
+	ctx := context.Background()
+	apiClient := testEnv.APIClient()
+
+	cID := container.Run(ctx, t, apiClient, func(c *container.TestContainerConfig) {
+		c.Config.Healthcheck = &containertypes.HealthConfig{
+			Test:     []string{"CMD-SHELL", `sh -c "sleep 60"`},
```

Comment:
Relevant CI logs: [builtin](https://github.com/moby/moby/runs/7949749461?check_suite_focus=true#step:21:2952), [`containerd-shim-runhcs-v1`](https://github.com/moby/moby/runs/7949748914?check_suite_focus=true#step:19:2368)

In both cases, the `CommandLine` string ultimately passed into the Win32API `CreateProcess` is
```
cmd /S /C "sh -c \"sleep 60\""
```

After lots of research and experimentation on a Windows box... my head hurts. Here's what I know so far:
- A Windows process' command and arguments are passed in as a single string, which is the string specified in the `CommandLine` passed into `CreateProcess`.
- Each process parses its own command line. Not all processes use the same algorithm.
  - As a corollary,`cmd.exe` does not tokenize its subprocess' command lines for them.
- `cmd.exe /S /C` follows some arcane rules, but the gist is that everything following the `/C` is taken literally as a command line, with one exception: if the first non-whitespace character following the `/C` is a double-quote character, the command line is modified by dropping that character and the last double-quote character in the remaining command-line.

Putting that all together:
`cmd /S /C "sh -c \"sleep 60\""` is equivalent to interactively running
```
C:\> sh -c \"sleep 60\"
```
Executing that command results in `CreateProcess()` being called with `CommandLine` set to `sh -c \"sleep 60\"`.

The Windows C Runtime parses the `CommandLine` `sh -c \"sleep 60\"` into the argv
```
argv[0] = sh
argv[1] = -c
argv[2] = \"sleep
argv[3] = 60\"
```
and busybox ash gets very confused.

The ultimate issue is that `cmdProbe` is incorrectly escaping the shell command line in a manner which the Windows C runtime would parse as a single argument, expecting `cmd.exe /S /C` to act like `sh -c` and consume the next argv as its argument. But `cmd.exe` does not unescape the argument.

https://github.com/containerd/containerd/issues/5067 is very related to the above, but I don't think it has to do with the timeouts you saw. (My guess is that the timeouts have something to do with `sleep.exe` not existing.)

Reply:
I recall I tested various combinations; any of them would fail on at least _one_ combination 😞 

I tried running `sleep` as main process itself;

```
[]string{"CMD", "sleep",  "60"},
```

Which would be the equivalent of the default command we set for "sleep" containers; https://github.com/moby/moby/blob/7b9275c0da707b030e62c96b679a976f31f929d3/integration-cli/test_vars_test.go#L6-L9

Tried running it as child process of `sh`;

```
[]string{"CMD", "sh", "-c", "sleep 60"},
```

Tried passing it as single string, but I think Linux wasn't a fan of that

```
[]string{"CMD-SHELL", `sh -c "sleep 60`},
```


And the original one (which (like above), and if I'm not mistaken, would run `sh` as child process of `cmd`, so that didn't feel right);

```
[]string{"CMD-SHELL", "sh", "-c", "sleep 60"},
```

Reply:
One thing I wonder is if the Busybox image should also be updated to set the `SHELL` to `/bin/sh`; currently it doesn't, which means that (in Dockerfiles), the default shell would still be `cmd` or `powershell` (depending on the base image)

Reply:
Also wondering if `exec.Config` needs to have a `CommandLine` property added; it just occurred to me that we have that for the container's main process, but don't have it for `exec` 🤔 https://github.com/moby/moby/blob/2646beac875eaf69930387d04fceca89e93d1da1/vendor/github.com/opencontainers/runtime-spec/specs-go/config.go#L41-L43

Reply:
> One thing I wonder is if the Busybox image should also be updated to set the `SHELL` to `/bin/sh`

> Also wondering if `exec.Config` needs to have a `CommandLine` property added

Both excellent ideas I wholeheartedly support!